router/tests/extract_routes.rs

```diff
@@ -0,0 +1,167 @@
+#![cfg(feature = "ssr")]
+
+use itertools::Itertools;
+use leptos::*;
+use leptos_router::{
+    generate_route_list_inner, Branch, Route, Router, Routes, TrailingSlash,
+};
+
+#[component]
+fn DefaultApp() -> impl IntoView {
+    let view = || view! { "" };
+    view! {
+        <Router>
+            <Routes>
+                <Route path="/foo" view/>
+                <Route path="/bar/" view/>
+                <Route path="/baz/:id" view/>
+                <Route path="/baz/:name/" view/>
+                <Route path="/baz/*any" view/>
+            </Routes>
+        </Router>
+    }
+}
+
+#[component]
+fn ExactApp() -> impl IntoView {
+    let view = || view! { "" };
+    let trailing_slash = TrailingSlash::Exact;
+    view! {
+        <Router trailing_slash>
+            <Routes>
+                <Route path="/foo" view/>
+                <Route path="/bar/" view/>
+                <Route path="/baz/:id" view/>
+                <Route path="/baz/:name/" view/>
+                <Route path="/baz/*any" view/>
+            </Routes>
+        </Router>
+    }
+}
+
+#[component]
+fn RedirectApp() -> impl IntoView {
+    let view = || view! { "" };
+    let trailing_slash = TrailingSlash::Redirect;
+    view! {
+        <Router trailing_slash>
+            <Routes>
+                <Route path="/foo" view/>
+                <Route path="/bar/" view/>
+                <Route path="/baz/:id" view/>
+                <Route path="/baz/:name/" view/>
+                <Route path="/baz/*any" view/>
+            </Routes>
+        </Router>
+    }
+}
+#[test]
+fn test_generated_routes_default() {
+    // By default, we use the behavior as of Leptos 0.5, which is equivalent to TrailingSlash::Drop.
+    assert_generated_paths(
+        DefaultApp,
+        &["/bar", "/baz/*any", "/baz/:id", "/baz/:name", "/foo"],
+    );
+}
+
+#[test]
+fn test_generated_routes_exact() {
+    // Allow users to precisely define whether slashes are present:
+    assert_generated_paths(
+        ExactApp,
+        &["/bar/", "/baz/*any", "/baz/:id", "/baz/:name/", "/foo"],
+    );
+}
+
+#[test]
+fn test_generated_routes_redirect() {
+    // TralingSlashes::Redirect generates paths to redirect to the path with the "correct" trailing slash ending (or lack thereof).
+    assert_generated_paths(
+        RedirectApp,
+        &[
+            "/bar",
+            "/bar/",
+            "/baz/*any",
+            "/baz/:id",
+            "/baz/:id/",
+            "/baz/:name",
+            "/baz/:name/",
+            "/foo",
+            "/foo/",
+        ],
+    )
+
+    // TODO:
+    // Test we get a redirect from "/foo/" to "/foo"
+    // Test we get a redirect from "/bar" to "/bar/".
```

Comment:
I've tested this locally in my browser, but I'm not immediately sure how to test it in a unit test. Suggestions welcome!  😊